src/ucp/stream/stream_recv.c

```diff
@@ -30,110 +62,311 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_stream_recv_data_nb,
+            goto out;
+        }
+    }
+
+    ucs_assert(req->recv.state.offset == req->recv.length);
+    req = ucp_stream_recv_request_put(req, count, UCS_OK);
+
+out:
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
+    return UCS_PTR_IS_PTR(req) ? (req + 1) : req;
+}
+
+static UCS_F_ALWAYS_INLINE ucp_recv_desc_t *
+ucp_stream_am_rdesc_get(ucp_worker_t *worker, void *data, size_t length,
+                        unsigned am_flags)
+{
+    ucp_recv_desc_t *rdesc;
 
     if (ucs_unlikely(am_flags & UCT_CB_PARAM_FLAG_DESC)) {
         /* slowpath */
         rdesc        = (ucp_recv_desc_t *)data - 1;
         rdesc->flags = UCP_RECV_DESC_FLAG_UCT_DESC;
-        status       = UCS_INPROGRESS;
     } else {
-        status = UCS_OK;
+        /* TODO: optimize if there are posted recvs */
         rdesc = (ucp_recv_desc_t*)ucs_mpool_get_inline(&worker->am_mp);
         if (ucs_likely(rdesc != NULL)) {
             rdesc->flags = 0;
             memcpy(rdesc + 1, data, length);
         } else {
-            ucs_error("ucp recv descriptor is not allocated");
-            goto out;
+            ucs_fatal("ucp recv descriptor is not allocated");
         }
     }
 
-    rdesc->length  = recv_len;
-    rdesc->hdr_len = hdr_len;
-    hdr            = data;
+    rdesc->length  = length - sizeof(ucp_stream_am_hdr_t);
+    rdesc->hdr_len = sizeof(ucp_stream_am_hdr_t);
+
+    return rdesc;
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_stream_ep_enqueue(ucp_ep_ext_stream_t *ep, ucp_worker_h worker)
+{
+    ucs_list_add_tail(&worker->stream_eps, &ep->list);
+    ep->ucp_ep->flags |= UCP_EP_FLAG_STREAM_IS_QUEUED;
+}
 
-    hash_it = kh_get(ucp_worker_ep_hash, &worker->ep_hash, hdr->sender_uuid);
-    if (ucs_unlikely(hash_it == kh_end(&worker->ep_hash))) {
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_stream_am_handler(void *am_arg, void *am_data, size_t am_length,
+                      unsigned am_flags)
+{
+    ucp_worker_h        worker      = am_arg;
+    ucp_stream_am_hdr_t *hdr        = am_data;
+    ucp_ep_ext_stream_t *ep_stream  = NULL;
+    ucp_ep_h            ep;
+    ucp_recv_desc_t     *rdesc;
+    ucp_request_t       *req;
+    ssize_t             unpacked;
+
+    ucs_assert(am_length >= sizeof(ucp_stream_am_hdr_t));
+
+    rdesc = ucp_stream_am_rdesc_get(worker, am_data, am_length, am_flags);
+    ucs_assert(rdesc != NULL);
+
+    ep = ucp_worker_ep_find(worker, hdr->sender_uuid);
+    if (ucs_unlikely(ep == NULL)) {
         ucs_error("ep is not found by uuid: %lu", hdr->sender_uuid);
-        goto err;
+        goto out;
     }
 
-    ep = kh_value(&worker->ep_hash, hash_it);
+    ep_stream = ep->ext.stream;
 
-    if (!(ep->flags & UCP_EP_FLAG_STREAM_IS_QUEUED)) {
-        ep->flags |= UCP_EP_FLAG_STREAM_IS_QUEUED;
-        ucs_list_add_tail(&worker->stream_eps, &ep->ext.stream->list);
+    if (ep->flags & UCP_EP_FLAG_STREAM_IS_QUEUED) {
+        ucs_queue_push(&ep_stream->data, &rdesc->stream_queue);
```

Comment:
pls optimize - need to push to a queue only if the data cannot be fully unpacked into the outstanding "expected" requests